virtool/analyses/api.py

```diff
@@ -174,6 +186,93 @@ async def remove(req: aiohttp.web.Request) -> aiohttp.web.Response:
     return no_content()
 
 
+@routes.post("/api/analyses/{id}/files", permission="upload_file")
+async def upload(req: aiohttp.web.Request) -> aiohttp.web.Response:
+    """
+    Upload a new analysis result file to the `analysis_files` SQL table and the `analyses` folder in the Virtool
+    data path.
+
+    """
+    db = req.app["db"]
+    pg = req.app["pg"]
+    analysis_id = req.match_info["id"]
+    analysis_format = req.query.get("format")
+
+    document = await db.analyses.find_one(analysis_id)
+
+    if document is None:
+        return not_found()
+
+    errors = virtool.uploads.utils.naive_validator(req)
+
+    if errors:
+        return invalid_query(errors)
+
+    name = req.query.get("name")
+
+    if analysis_format not in ANALYSIS_FORMATS:
+        return bad_request("Unsupported analysis file format")
+
+    analysis_file = await virtool.analyses.db.create_row(pg, analysis_id, analysis_format, name)
+
+    name_on_disk = analysis_file["name_on_disk"]
+    files = document.get("files", [])
+
+    if name_on_disk in files:
+        return conflict("File is already associated with analysis")
+
+    file_id = analysis_file["id"]
+    analysis_file_path = Path(req.app["settings"]["data_path"]) / "analyses" / name_on_disk
+
+    try:
+        size = await virtool.uploads.utils.naive_writer(req, analysis_file_path)
+    except asyncio.CancelledError:
+        logger.debug(f"Analysis file upload aborted: {file_id}")
+        await virtool.analyses.db.delete_row(pg, file_id)
+
+        return aiohttp.web.Response(status=499)
+
+    analysis_file = await virtool.uploads.db.finalize(pg, size, file_id, AnalysisFile)
+
+    if not analysis_file:
+        await req.app["run_in_thread"](os.remove, analysis_file_path)
+        return not_found("Row not found in table after file upload")
+
+    files.append(name_on_disk)
+
+    await db.analyses.update_one({"_id": analysis_id}, {
+        "$set": {"files": files}
+    })
```

Comment:
Use `$push`. If there are two requests being handled at the same time (ie. the workflow is concurrently uploading two files), rewriting the `files` field could cause a serious bug.